### DISCUSSION TITLE
what's the best way to detect a python type in rust?

### DISCUSSION THREAD

**Discussion by shi-yan** [PyO3/pyo3#1914] [Questions]
I'm trying to serialize a python dict into a json in rust.

what's the best way to detect python types?

currently I'm doing python_var.get_type().name().unwrap(), with a string comparison. it seems  to be inefficient?

is there a better way doing this? Thanks!

Another question, how to initialize python's NoneType? there is no document covering that?

And how can I convert from PyDict to PyAny? 

Thanks.

**Comment by birkenfeld** [comment]
You should be able to directly compare the result of `get_type()?` (better not use unwrap unless forced) for equality.  However, `obj.is_instance` is also available which might be a better fit.

What do you mean by "initialize NoneType"?

`&PyDict` should automatically deref to `&PyAny`, so for specific problems please post a bit of code.

**Reply by shi-yan** [reply]
Thank you.

I'm creating a binding for a no-sql database. I need to convert what the db returns into a python object.

The database uses json-like format. an entry can be a dict-like {'test':32} or it can be an array like [23,23,23...]

therefore, I want my function return PyAny.
```rust
fn fetch_from_db(...) -> PyResult<PyAny> {

if (db.result.is_array) {
create a PyList and return
}
else is_dict {
create a PyDict and return
}
else is_null {
create a python equivalent of None in rust and return.
}
}
```
so NoneType is python's equivalent of None , because null is valid for json.

&PyAny can't be returned.

**Reply by mejrs** [reply]
You need to return the owned version of `Py<PyAny>` (that is what the `PyObject` alias is), if lifetimes forbid you from returning the borrowed variant. See also https://pyo3.rs/main/doc/pyo3/conversion/trait.IntoPy.html#dynamic-conversion-into-python-objects

**Reply by birkenfeld** [reply]
You should still be able to return `&PyAny` if you associate it with an incoming `'py` lifetime, i.e.
```
fn fetch<'py>(py: Python<'py>, ...) -> PyResult<&'py PyAny> { ... }
```
But `Py<PyAny>` is just as fine, don't worry too much about that.  The most important thing is that you can convert between most of the different object types in some way or other, [which is described in detail here](https://pyo3.rs/v0.14.5/types.html#object-types).

**Comment by davidhewitt** [comment]
By the way, from what you're doing sounds like https://github.com/davidhewitt/pythonize might be useful?
